### PR TITLE
New: Inform when fixes were applied

### PR DESCRIPTION
**Pull request by pfhayes**
**What is the purpose of this pull request? (put an "X" next to item)**
[X] Add something to the core

**What changes did you make? (Give an overview)**
Emits informational text that fixes were applied.

I've updated the `stylish` formatter to emit text when the source files have been changed due to autofixes.

Since it is difficult to detect how many fixes were applied, we instead just emit "Fixes were applied.". This can indicate to the viewer that they should use `git status` or some other tool to inspect the changes further.

Merging this in will be a breaking change for users who expect the stylish formatter to not output any text on a 0 exit code.

### CODE REVIEW

File: tests/bin/eslint.js [hunk 1]
```diff
@@ -127,6 +127,22 @@ describe("bin/eslint.js", () => {
...
+        it("lets you know that files were fixed", () => {
+            const child = runESLint(["--fix", "--no-eslintrc", "--no-ignore", tempFilePath]);
+
+            return getStdout(child).then(stdout => {
+                assert.match(stdout, /Fixes were applied/);
+            });
+        });
+
+        it("won't falsely tell you that files were fixed", () => {
+            const child = runESLint(["--fix", "--no-eslintrc", "--no-ignore", expectedFixedText]);
```
Comment (not-an-aardvark) [inline tests/bin/eslint.js]: This line doesn't seem right to me; it's running ESLint with the fixed text as an argument. I think you meant to use the *filepath* of the fixed text instead.

**Comment by gyandeeps** [comment]
I dont think the current approach is the right approach to the problem. I think we need to go back to the drawing board (on the issue) to discuss this further.
Thanks @pfhayes for your efforts. Lets work through the API level solution as I mentioned in the issue.

**Comment by nzakas** [comment]
Agree with @gyandeeps , I don't believe this is the right approach for this problem.

**Comment by pfhayes** [comment]
PR has been updated to only update the `stylish` formatter

File: tests/bin/eslint.js (hunk 1 above)
Comment (pfhayes) [inline tests/bin/eslint.js]: Good catch, you are correct. this test has been removed in the most recent iteration, however

**Comment by pfhayes** [comment]
I updated the PR title to address the comment from @eslintbot - I assume that if the PR is squashed/rebased in that will address the issue

**Comment by nzakas** [comment]
Just a heads up: we have moved to a new CLA checker on pull requests. Even if you've previously signed our CLA, we will need to you sign the new one. To do so, look at the status checks for **licence/cla** and click the "Details" link. Sorry for the inconvenience.

**Comment by pfhayes** [comment]
thanks @nzakas - I have resigned

File: lib/formatters/stylish.js [hunk 2]
```diff
@@ -21,6 +21,16 @@ function pluralize(word, count) {
     return (count === 1 ? word : `${word}s`);
 }
 
+/**
+ * Returns true iff any fixes were applied to any inpout files
```
Comment (vitorbal) [inline lib/formatters/stylish.js]: typo: `iff`
typo: `inpout`

Comment (pfhayes) [inline lib/formatters/stylish.js]: Thanks @vitorbal - I've updated this comment

**Comment by pfhayes** [comment]
I have force-pushed to update the PR to be a single commit with the tag "New:", but @eslintbot is still requesting changes. Is there something else I need to change?

**Comment by vitorbal** [comment]
Thanks! Don't worry about the commit message. We can always fix it whenever
we squash and merge this.
For now, let's wait to see what the rest of the team thinks about the
concerns raised :-)

**Comment by kaicataldo** [comment]
Marked as accepted as per the TSC resolution. Needs discussion as to what the actual implementation will be.

**Comment by vitorbal** [comment]
@eslint/eslint-team: Let's try to get the discussion going for this one again. Its purpose is to emit informational text that fixes were applied. Right now this PR implements this only for the output of the stylish formatter. Here's what the TSC resolution was:

> More discussion is needed for both issues in regards to formatter changes. We should try to make them non-breaking, and if we can't do that, we should make a breaking change that would support other cases (like deprecated rules)

In my opinion, changing the output of the stylish and codeframe formatters shouldn't be considered a breaking change, but I don't think everyone agrees. I'm having a hard time otherwise, thinking of a way to make this change non-breaking. Any ideas?

**Comment by mikesherov** [comment]
Just catching up here. Do these formatters emit text when no errors were found? And does ESLint emit a non-zero exit code if only fixed errors were present?

I ask because if ESLint emits a 0 and now suddenly also emits text, it might break the old Linux design principle of "Rule of Silence": http://www.linfo.org/rule_of_silence.html With that said, the theoretical gain in UX is offset by the risk that systems assume eslint acts like other cli tools that respect the rule of silence.

Seems if we're definitely moving forward with this change, the only guaranteed BC way to do this is create new reporters: codeframe-noisy and stylish-noisy. Feel free to bikeshed the names ;-)

**Comment by nzakas** [comment]
> In my opinion, changing the output of the stylish and codeframe formatters shouldn't be considered a breaking change, but I don't think everyone agrees

Unfortunately, almost any change to formatters is breaking because we can't know if someone is relying on the format to be automatically parsed. 

The only non breaking change would be to create a new formatter that includes fix information.

**Comment by nzakas** [comment]
> Do these formatters emit text when no errors were found?

Formatters are not currently run when no errors are found. 

> And does ESLint emit a non-zero exit code if only fixed errors were present?

It's a nonzero exit code when there are no problems to report. So if auto fix fixed every error, thus resulting in zero reported errors, the exit code is 0.

**Comment by ilyavolodin** [comment]
> Unfortunately, almost any change to formatters is breaking because we can't know if someone is relying on the format to be automatically parsed.
The only non breaking change would be to create a new formatter that includes fix information.

I think this is a very limiting position for us, especially when it comes to human-readable formatters. I completely agree on this position when it comes to `JSON`, `UNIX`, `tap` or other machine-readable formatters, but I think `stylish`, `codeframe`, `HTML` and `table` should be excluded from this.

**Comment by nzakas** [comment]
> I think this is a very limiting position for us, especially when it comes to human-readable formatters.

We can't decide which formatters are ok to change after the fact. We've never communicated that to users, and I don't think it's fair for us to do so arbitrarily for this purpose. I know it's limiting, but that is the cost of maintaining a popular tool. We are deep in people's builds and can't assume anything. 

I do think we should do a major release soon, so another option is to make this a breaking change and communicate it accordingly.

**Comment by vitorbal** [comment]
I think making this change part of a major release and communicating it
clearly is a good option.

**Comment by pfhayes** [comment]
Thanks everyone - I agree that this makes sense as a breaking change and should go out with a major version increase.

I've pushed a fix for the previous build breakage - please let me know if there is anything else I can do for this PR

**Comment by nzakas** [comment]
@pfhayes can you please update the description of this PR to explain the approach you've taken and what the consequences of merging this in will be?

**Comment by pfhayes** [comment]
@nzakas sure - I've updated the description to include more detailed information

**Comment by pfhayes** [comment]
Anything else I can provide here? cc @not-an-aardvark

File: lib/formatters/stylish.js [hunk 3]
```diff
@@ -21,6 +21,16 @@ function pluralize(word, count) {
...
 }
 
+/**
+ * Returns true if and only if any fixes were applied to any input files
+ * @param {Object[]} results - Collection of messages from all the files
+ * @returns {boolean} If any fixes were applied
+ * @private
+ */
+function fixesWereApplied(results) {
+    return Boolean(results && results.some(r => r.output));
```
Comment (not-an-aardvark) [inline lib/formatters/stylish.js]: Is `results && ...` necessary here? From what I can tell, `results` will always be truthy.

Comment (pfhayes) [inline lib/formatters/stylish.js]: Ok - I've removed this extraneous check

**Comment by not-an-aardvark** [comment]
I updated the labels based on the discussion in https://github.com/eslint/eslint/issues/7404#issuecomment-336541458.

**Comment by not-an-aardvark** [comment]
Closing because https://github.com/eslint/eslint/issues/7404 was closed.

**Pull request closed without merging**